llvm/lib/RemoteCachingService/GRPCRelayCAS.cpp

```diff
@@ -0,0 +1,638 @@
+  CASPutResponse Response;
+  Request.mutable_data()->CopyFrom(createCASData(None, ""));
+  grpc::Status Status = CASStub->Put(&Context, Request, &Response);
+  if (!Status.ok()) {
+    Err = createGRPCStatusError(Status);
+    return;
+  }
+  if (Response.has_error()) {
+    Err = createGRPCResponseError(Response.error());
+    return;
+  }
+  ServiceHashSize = Response.cas_id().id().size();
+  if (ServiceHashSize > 80)
+    Err = createStringError(std::make_error_code(std::errc::message_size),
+                            "Hash from CASService is too long");
+}
+
+StringRef GRPCRelayCAS::getHashSchemaIdentifier() const {
+  return "grpc::cas::service::v1";
+}
+
+ArrayRef<uint8_t> GRPCRelayCAS::getHashImpl(const CASID &ID) const {
+  ArrayRef<uint8_t> ExtendedHash(extractIndexFromID(ID).Hash);
+  return ExtendedHash.take_front(ServiceHashSize);
+}
+
+// FIXME: this is a partial implementation of current CAS implementation.
+// Should move this into GRPC protocol.
+static StringRef getCASIDPrefix() { return "0~"; }
+
+static std::string decodeBase64(std::string const &encoded_string) {
+  size_t in_len = encoded_string.size();
+  size_t i = 0;
+  size_t j = 0;
+  int in_ = 0;
+  unsigned char char_array_4[4], char_array_3[3];
+  std::string ret;
+
+  auto is_base64 = [](unsigned char c) {
+    return (isalnum(c) || (c == '+') || (c == '/'));
+  };
+
+  const std::string base64_chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+                                   "abcdefghijklmnopqrstuvwxyz"
+                                   "0123456789+/";
+
+  while (in_len-- && (encoded_string[in_] != '=') &&
+         is_base64(encoded_string[in_])) {
+    char_array_4[i++] = encoded_string[in_];
+    in_++;
+    if (i == 4) {
+      for (i = 0; i < 4; i++)
+        char_array_4[i] =
+            static_cast<unsigned char>(base64_chars.find(char_array_4[i]));
+
+      char_array_3[0] =
+          (char_array_4[0] << 2) + ((char_array_4[1] & 0x30) >> 4);
+      char_array_3[1] =
+          ((char_array_4[1] & 0xf) << 4) + ((char_array_4[2] & 0x3c) >> 2);
+      char_array_3[2] = ((char_array_4[2] & 0x3) << 6) + char_array_4[3];
+
+      for (i = 0; (i < 3); i++)
+        ret += char_array_3[i];
+      i = 0;
+    }
+  }
+
+  if (i) {
+    for (j = i; j < 4; j++)
+      char_array_4[j] = 0;
+
+    for (j = 0; j < 4; j++)
+      char_array_4[j] =
+          static_cast<unsigned char>(base64_chars.find(char_array_4[j]));
+
+    char_array_3[0] = (char_array_4[0] << 2) + ((char_array_4[1] & 0x30) >> 4);
+    char_array_3[1] =
+        ((char_array_4[1] & 0xf) << 4) + ((char_array_4[2] & 0x3c) >> 2);
+    char_array_3[2] = ((char_array_4[2] & 0x3) << 6) + char_array_4[3];
+
+    for (j = 0; (j < i - 1); j++)
+      ret += char_array_3[j];
+  }
+
+  return ret;
+}
+
+void GRPCRelayCAS::printIDImpl(raw_ostream &OS, const CASID &ID) const {
+  assert(&ID.getContext() == this);
+  ArrayRef<uint8_t> Hash = ID.getHash();
+  assert(Hash[0] == 0 && "Expected to start with 0");
```

Comment:
I believe this constraint will be found rather inconvenient at some point, as the schemas evolve.